readthedocs/doc_builder/builder.py

```diff
@@ -0,0 +1,443 @@
+import os
+from collections import defaultdict
+
+import structlog
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
+
+from readthedocs.builds.constants import EXTERNAL
+from readthedocs.doc_builder.config import load_yaml_config
+from readthedocs.doc_builder.loader import get_builder_class
+from readthedocs.doc_builder.python_environments import Conda, Virtualenv
+from readthedocs.projects.exceptions import RepositoryError
+from readthedocs.projects.models import Feature
+from readthedocs.projects.signals import after_build, before_build, before_vcs
+
+log = structlog.get_logger(__name__)
+
+
+class DocumentationBuilder:
+    def __init__(
+        self,
+        data,  # Celery task's ``data`` object
+    ):
+        """Initializer.
+
+        :param data: object with all the data grabbed by Celery task in
+        ``before_start`` and used as a way to share data with this class
+        by-directionally.
+
+        :type data: readthedocs.projects.tasks.builds.TaskData
+
+        """
+        self.data = data
+
+    def vcs(self):
```

Comment:
This needs a better name. `sync_vcs` or similar?

Reply:
Maybe `vcs_checkout` / `vcs_clone`? It seems to be "checkout" the main goal of this method. However, I'm not sure if I want to name it similar to the other `_checkout` methods. I think that's why I called it just `vcs` because it encapsulates the checkout and other VCS related tasks (load config and checkout submodules).

Reply:
`setup_vcs` could be, I think that name is used in other places, or `setup_project/repo`.

Reply:
Just something a bit more descriptive. It needs a verb to be more clear, I don't particularly care what that verb is.